### DISCUSSION TITLE
Publish ros2_rust as a crate on crates.io

### DISCUSSION THREAD

**Discussion by keenanjohnson** [ros2-rust/ros2_rust#86] [General]
Hey all!

One thing that would really make the use of this library easier is if it was deployed as a crate on cargo.io.

I'm guessing there are probably some technical hurdles left before this is possible, but I would like to help make this happen if possible? 

Do any of the maintainers have a sense of what would need to get done before this is possible?

Thanks for the great work!

Related https://github.com/ros2-rust/ros2_rust/issues/9

**Comment by esteve** [comment]
The current owner of the `rclrs` crate on `crates.io` has agreed to transfer it. However, even if we publish `rclrs` on crates.io, it won't be as straightforward as with any other crate, because we need some of the ROS2 infrastructure and message packages still need to be built via `colcon` for now.

**Reply by keenanjohnson** [reply]
Yeah definitely understood that it's not as simple as a cargo import, but I think the cargo module would be much better than the current approach I have which is importing the git repo as a submodule as seen here for example: https://github.com/keenanjohnson/ros2_rust_workspace/tree/main/src

**Comment by keenanjohnson** [comment]
Just checking in on the status of this?

**Comment by esteve** [comment]
We want the first release to include services (https://github.com/ros2-rust/ros2_rust/pull/146), so that users have a minimum set of features (pubsub, rpc) for developing applications, after that we'll be ready to release ros2-rust as a crate.

**Comment by keenanjohnson** [comment]
Sounds great! Looks like you have done some awesome work in that PR @esteve ! Let us know if you need any additional support there to close that out!

**Comment by esteve** [comment]
@keenanjohnson yeah, give the PR a try if you have a moment. I estimate I'll tag `ros2-rust` and release the crate next week, and I'll make the announcement on Discourse. However, not everything will be released on crates.io, the code generator for example and all the build tools will be released separately.

**Comment by esteve** [comment]
@keenanjohnson I've updated the description of this discussion to point to crates.io (instead of cargo.io). We had a ticket open (https://github.com/ros2-rust/ros2_rust/issues/9) a while ago that I've linked to.

**Comment by keenanjohnson** [comment]
Hey all! I see that #146 was merged a while ago. Is the plan still to release on crates.io soon or are there other blockers here?

**Reply by nnmm** [reply]
We did publish the last two releases to crates.io: https://crates.io/crates/rclrs

**Comment by keenanjohnson** [comment]
Oh excellent! Are there specific instructions anywhere for installing via
cargo? I had looked at the readme but didn’t see any so I had assumed it
wasn’t publish yet. I know in the chat channel at some point it was
discussed that there would still be additional steps after the standard
cargo install command to get it to work.

**Reply by nnmm** [reply]
`cargo install` is only for binaries, so you only need to write `rclrs = 0.3` in your `Cargo.toml` and have a ROS installation sourced.

If you want to use message packages however, you additionally need to 

1. Add to your workspace
   - `rosidl_generator_rs` from `ros2_rust`
   - `https://github.com/ros2/rosidl_core.git` and `https://github.com/ros2/rosidl_defaults.git`
   - The message packages you want to generate Rust code for
2. Add the proper dependency to your `Cargo.toml`, e.g. `geometry_msgs = "*"`
3. Run `colcon build` to generate the Rust package and set up `.cargo/config.toml`

I should check if the build instructions are still up to date.

**Comment by esteve** [comment]
[rclrs](https://crates.io/crates/rclrs) and [rosidl_runtime_rs](https://crates.io/crates/rosidl_runtime_rs) are now on crates.io
